servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java

```diff
@@ -0,0 +1,134 @@
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.worker.block.evictor;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+
+import tachyon.Constants;
+import tachyon.Pair;
+import tachyon.worker.BlockStoreLocation;
+import tachyon.worker.block.BlockAccessEventListener;
+import tachyon.worker.block.BlockMetadataManager;
+import tachyon.worker.block.meta.BlockMeta;
+
+public class LRUEvictor implements Evictor, BlockAccessEventListener {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private final BlockMetadataManager mMeta;
+
+  /** Double-Link List, most recently accessed block is at tail of the list */
+  private class Node extends DoubleLinkListNode {
+    long blockId;
+
+    Node() {
+      blockId = -1;
+      prev = next = null;
+    }
+
+    Node(long blkId) {
+      blockId = blkId;
+      prev = next = null;
+    }
+
+    Node nextNode() {
+      return (Node) next;
+    }
+  }
+
+  private Node mHead;
+  private Node mTail;
+  /** Map from blockId to corresponding Node in the Double-Link List */
+  private Map<Long, Node> mCache;
+
+
+  public LRUEvictor(BlockMetadataManager meta) {
+    mMeta = meta;
+
+    mHead = new Node();
+    mTail = new Node();
+    mHead.append(mTail);
+    mCache = new HashMap<Long, Node>();
+  }
+
+  /**
+   * Free space in the given block store location. The location can be a specific location, or
+   * {@link BlockStoreLocation#anyTier()} or {@link BlockStoreLocation#anyDirInTier(int)} .
+   *
+   * If the total free space is fewer than {@code bytes}, they will all be evicted
+   * 
+   * Thread safe.
+   *
+   * @param bytes the size in bytes
+   * @param location the location in block store
+   * @return an eviction plan to achieve the freed space
+   */
+  @Override
+  public Optional<EvictionPlan> freeSpace(long bytes, BlockStoreLocation location) {
+    List<Pair<Long, BlockStoreLocation>> toMove = new ArrayList<Pair<Long, BlockStoreLocation>>();
+    List<Long> toEvict = new ArrayList<Long>();
+
+    Node p = mHead.nextNode();
+    long evictBytes = 0;
+    // erase race condition with onAccessBlock on internal data structure
+    synchronized (mTail) {
+      while (p != mTail && evictBytes < bytes) {
+        Optional<BlockMeta> meta = mMeta.getBlockMeta(p.blockId);
+        boolean evicted = false;
+        if (!meta.isPresent()) {
+          LOG.error("BlockMeta for blockId %d can not be get in LRUEvictor", p.blockId);
+        } else if (meta.get().getBlockLocation().belongTo(location)) {
```

Comment:
If the passed in block store location is anyDirInTier, I don't think this will work since this sums the total space available across all dirs?